### PR TITLE
Windows Notepad: report line numbers when that setting is on. #16354

### PR DESCRIPTION
**Pull request by tburc**
### Link to issue number:

https://github.com/nvaccess/nvda/issues/16354

### Summary of the issue:

Windows 11 Notepad previously didn't read line numbers when line numbers setting was turned on.

### Description of user facing changes

NVDA will now announce line numbers while users work in Windows Notepad, if the NVDA setting to report line numbers is on.

### Description of development approach

- Change the Notepad AppModule to use a RichEdit overlay class for the text edit area

### Testing strategy:

- Turn on the setting: "Preferences > Document Formatting > Line numbers"
- Make a notepad file in Windows 11 and add text
- Press the up and down arrows.  
- Observe that NVDA announces line numbers.
- Open an existing notepad file, and press the up and down arrows
- Observe that NVDA read the line numbers in this case as well.

### Known issues with pull request:

By using a different overlay class, this potentially changes a lot of behavior around Windows Notepad.  It's possible this will introduce unanticipated regressions.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

@coderabbitai summary

### CODE REVIEW

**Comment by josephsl** [comment]
Hi,

One major regression: spelling errors are no longer detected and announced by NVDA in Windows 11 Notepad. Also, I advise adding a what's new entry somewhere.

Thanks.

File: source/appModules/notepad.py [hunk 1]
```diff
@@ -41,6 +42,10 @@ def event_UIA_elementSelected(self, obj: NVDAObject, nextHandler: Callable[[], N
 				)
 			)
 		nextHandler()
+
```
Comment (josephsl) [inline source/appModules/notepad.py:45]: We usually use a blank (empty) line between method definitions - in this line, it is a single tab with nothing after that.

**Comment by josephsl** [comment]
Hi,

Regarding spelling error detection regression and overall approach: I think adding an overlay class is just touching the surface. Line numbers are calculated based on what's called a "text info", a way to work with text ranges in editable content. But that's not all: line numbers are exposed if the control itself supports this, and in some cases, there are ways to obtain line number just by asking the control itself; in case of default UIA object like Windows 11 Notepad's editor control, line numbers are not exposed to NVDA because there is no generic way of doing this for UIA text ranges (@MichaelDCurran, am I correct?); this info comes from greping NVDA source code looking for the text "reportLineNumber".

Tanks.

**Comment by TristanBurchett** [comment]
> Regarding spelling error detection regression and overall approach: I think adding an overlay class is just touching the surface. 

Hi,

I agree that the issue is definitely centered on TextInfo objects.  Tracing through the logic there (starting with getFormatFieldSpeech and walking up the call stack) was a little mind bending :)

During the debugging process, I came across the comment on the NVDAObject constructor, stating that it's the responsibility of the overlay class to create the relevant TextInfo objects. So I think if we want to intervene in the TextInfo objects that Notepad creates, then we will need an overlay class of some kind, regardless.

Eventually, I noticed that for Notepad++ NVDA does report line numbers. In looking at the difference between notepad.py and notepadPlusPlus.py, I saw that for Notepad++, we use a Scintilla overlay corresponding to the scintilla window class.  Regular notepad doesn't use scintilla, instead it uses a rich edit, so this led me to believe that using an overlay class would be the best course of action.

There are very few places in the codebase where the "line-number" attribute is set.  Scintilla is one, and RichEdit is another.  All the other places seem specific to apps like Word.  So this seemed like the most likely approach.

The spelling regression is surprising and unfortunate.  I can look into that.  It might be best to try to fix that for all rich edit overlays?  Because presumably nothing that uses the RichEdit overlay currently will be able to support spelling corrections correctly.

**Pull request closed without merging**